core/src/main/java/io/questdb/cairo/wal/WalTxnYieldEventsImpl.java

```diff
@@ -0,0 +1,133 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+public class WalTxnYieldEventsImpl implements WalTxnYieldEvents {
+
+    private final CharSequence dbRoot;
+    private final ConcurrentHashMap<TableToken, ObjList<YieldEvent>> events = new ConcurrentHashMap<>();
+    private final FilesFacade ff;
+    private final MillisecondClock millisecondClock;
+    private final long spinLockTimeout;
+    private final TxReader txReader;
+    private final YieldEventFactory yieldEventFactory;
+
+    public WalTxnYieldEventsImpl(@NotNull CairoConfiguration cairoConfiguration, @NotNull YieldEventFactory yieldEventFactory) {
+        this.dbRoot = cairoConfiguration.getRoot();
+        this.ff = cairoConfiguration.getFilesFacade();
+        this.txReader = new TxReader(ff);
+        this.millisecondClock = cairoConfiguration.getMillisecondClock();
+        this.spinLockTimeout = cairoConfiguration.getSpinLockTimeout();
+        this.yieldEventFactory = yieldEventFactory;
+    }
+
+    @Override
+    public void close() {
+        for (Map.Entry<TableToken, ObjList<YieldEvent>> entry : events.entrySet()) {
+            Misc.freeObjListAndClear(entry.getValue());
+        }
+    }
+
+    @Override
+    public @Nullable YieldEvent register(TableToken tableToken, long txn) {
+        Path tlPath = Path.PATH.get().of(dbRoot);
+        tlPath.trimTo(dbRoot.length()).concat(tableToken).concat(TableUtils.META_FILE_NAME).$();
+        if (ff.exists(tlPath)) {
+            long currentTxn = readSeqTxn(tableToken, tlPath);
+            if (currentTxn == -1 || currentTxn >= txn) {
+                // The txn is already visible or the table is dropped.
+                return null;
+            }
+
+            final YieldEvent yieldEvent = yieldEventFactory.newInstance();
+            events.compute(tableToken, (t, list) -> {
+                if (list == null) {
+                    list = new ObjList<>();
+                }
+                list.add(yieldEvent);
+                return list;
+            });
+
+            // Read the txn once again to make sure ApplyWal2TableJob sees our changes.
+            try {
+                currentTxn = readSeqTxn(tableToken, tlPath);
+            } catch (CairoException e) {
+                yieldEvent.close();
+                throw e;
+            }
+
+            if (currentTxn == -1 || currentTxn >= txn) {
+                // The txn is already visible or the table is dropped, so we close the event
+                // and pretend it never existed. The event will be triggered and closed one
+                // more time later either after takeRegisteredEvents() or by close().
+                yieldEvent.close();
+                return null;
+            }
+            return yieldEvent;
+        }
+        // The table is dropped.
+        return null;
+    }
+
+    @Override
+    public void takeRegisteredEvents(TableToken tableToken, ObjList<YieldEvent> dest) {
+        events.computeIfPresent(tableToken, (t, list) -> {
```

Comment:
implicit allocation

Reply:
Same as above.

Reply:
Fixed in fa493333ca3bb0045b8f5cef9fd9d9fa4afb96ae